flash/data/batch.py

```diff
@@ -0,0 +1,186 @@
+        self.apply_per_sample_transform = apply_per_sample_transform
+        self.stage = stage
+
+    def forward(self, samples: Sequence[Any]):
+        if self.apply_per_sample_transform:
+            samples = [self.per_sample_transform(sample) for sample in samples]
+            samples = type(samples)(samples)
+            samples = self.collate_fn(samples)
+        samples = self.per_batch_transform(samples)
+        return samples
+
+    def __str__(self) -> str:
+        repr_str = '_PreProcessor:'
+        repr_str += f'\n\t(per_sample_transform): {repr(self.per_sample_transform)}'
+        repr_str += f'\n\t(collate_fn): {repr(self.collate_fn)}'
+        repr_str += f'\n\t(per_batch_transform): {repr(self.per_batch_transform)}'
+        repr_str += f'\n\t(apply_per_sample_transform): {repr(self.apply_per_sample_transform)}'
+        repr_str += f'\n\t(stage): {repr(self.stage)}'
+        return repr_str
+
+
+class _PostProcessor(torch.nn.Module):
+    """
+        This class is used to encapsultate the following functions of a PostProcess Object:
+        Inside main process:
+            per_batch_transform: Function to transform a batch
+            per_sample_transform: Function to transform an individual sample
+            uncollate_fn: Function to split a batch into samples
+            per_sample_transform: Function to transform an individual sample
+            save_fn: Function to save all data
+            save_per_sample: Function to save an individual sample
+    """
+
+    def __init__(
+        self,
+        uncollate_fn: Callable,
+        per_batch_transform: Callable,
+        per_sample_transform: Callable,
+        save_fn: Optional[Callable] = None,
+        save_per_sample: bool = False
+    ):
+        super().__init__()
+        self.uncollate_fn = convert_to_modules(uncollate_fn)
+        self.per_batch_transform = convert_to_modules(per_batch_transform)
+        self.per_sample_transform = convert_to_modules(per_sample_transform)
+        self.save_fn = convert_to_modules(save_fn)
+        self.save_per_sample = convert_to_modules(save_per_sample)
+
+    def forward(self, batch: Sequence[Any]):
+        uncollated = self.uncollate_fn(self.per_batch_transform(batch))
+
+        final_preds = type(uncollated)([self.per_sample_transform(sample) for sample in uncollated])
+
+        if self.save_fn is not None:
+            if self.save_per_sample:
+                for pred in final_preds:
+                    self.save_fn(pred)
+            else:
+                self.save_fn(final_preds)
+        else:
+            return final_preds
+
+    def __str__(self) -> str:
+        repr_str = '_PostProcessor:'
+        repr_str += f'\n\t(per_batch_transform): {repr(self.per_batch_transform)}'
+        repr_str += f'\n\t(uncollate_fn): {repr(self.uncollate_fn)}'
+        repr_str += f'\n\t(per_sample_transform): {repr(self.per_sample_transform)}'
+
+        return repr_str
+
+
+def default_uncollate(batch: Any):
```

Comment:
pls add a docstring with a sentence what is it for...

Reply:
also adding doctest example would help to understand and covers the code, now codecov complains it is not covered...

Reply:
Done !